src/Symfony/Component/Security/Http/EventListener/CheckRequiredBadgesListener.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Http\Event\CheckPassportEvent;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @experimental in 5.3
+ */
+final class CheckRequiredBadgesListener implements EventSubscriberInterface
+{
+    private $requiredBadges = [];
+
+    public function __construct(array $requiredBadges)
+    {
+        $this->requiredBadges = $requiredBadges;
+    }
+
+    public function checkPassport(CheckPassportEvent $event): void
+    {
+        $passport = $event->getPassport();
+        $missing = [];
+        foreach ($this->requiredBadges as $requiredBadge) {
+            if (!$passport->hasBadge($requiredBadge)) {
+                $missing[] = $requiredBadge;
+            }
+        }
+
+        if ($missing) {
+            throw new BadCredentialsException(sprintf('Authentication failed; Some badges marked as required by the firewall config are not available on the passport: "%s".', implode('", "', $missing)));
+        }
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            CheckPassportEvent::class => ['checkPassport', -512],
```

Comment:
does -512 imply we expect new badges during the checking stage?

shouldnt passport be immutable in that regard?

Reply:
They aren't, e.g. `UserProviderListener` automatically adds the password upgrade badge if your user provider supports upgrading passwords (and there are password credentials).